engine/src/components/component.h

```diff
@@ -0,0 +1,92 @@
+/*
+ * component.h
+ *
+ * Copyright (c) 2001-2002 Daniel Horn
+ * Copyright (c) 2002-2019 pyramid3d and other Vega Strike Contributors
+ * Copyright (c) 2019-2023 Stephen G. Tuggy, Benjamen R. Meyer, Roy Falk and other Vega Strike Contributors
+ *
+ * https://github.com/vegastrike/Vega-Strike-Engine-Source
+ *
+ * This file is part of Vega Strike.
+ *
+ * Vega Strike is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * Vega Strike is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Vega Strike. If not, see <https://www.gnu.org/licenses/>.
+ */
+
+// -*- mode: c++; c-basic-offset: 4; indent-tabs-mode: nil -*-
+
+#ifndef COMPONENT_H
+#define COMPONENT_H
+
+#include <string>
+#include <map>
+
+/**
+ * LibComponent is currently tightly coupled to LibDamage and
+ * other various libraries in VegaStrike engine.
+ * Consider decoupling and subclassing every component in it.
+ */
+
+class Unit;
+
+// TODO: add complete list
+enum class ComponentType {
+    Hull, 
+    Armor, 
+    Shield,
+    Drive
+};
+
+class Component
+{
+protected:
```

Comment:
just to note, we should put `public` information first, then `protected`, then `private`. This helps with ABI compatibility, not that we're really concerned about that right now.

Reply:
For some reason I assumed the reverse. I'll fix this here, though I can't promise for all of my work.

Reply:
Changing things changes the offsets. Keeping public data first keeps the offsets to it the same, and new public goes at the end, pushing the protected and private data further away from the zero point.

If we really wanted a public binary ABI then we'd look at doing another method (what Qt does), but it's not really important just good habit for when it is.

Reply:
So, I dug deeper and while most agree on public/protected/private order, fields/constructors/methods is less so.
Since most of the code is fields/constructors/methods, I'll stick with that but reverse the visibility as I go along as these are already not consistent.